Move GameLoop into a thread (and no longer run Paint in a thread)

TrueBrain opened the pull request:
Requires #8747 to be merged first.

## Motivation / Problem

During work on the video drivers, I noticed a few (SDL1, SDL2 and Win32) had a "draw thread" implementation, where drawing was moved to a thread. This to me is odd, as normally you want to do drawing in the main thread, and everything else in other threads.

This was also showing, as with OpenGL, "draw thread" could not be used. The thread the context is created in, has to be the thread that is updating it. And given this was done in the main thread, it could not be done in the "draw thread".

This PR introduced an alternative: do `GameLoop` in a thread. It works for all video drivers.

## Description

```
This allows drawing to happen while the GameLoop is doing an
iteration too.
    
Sadly, not much drawing currently can be done while the GameLoop
is running, as for example PollEvent() or UpdateWindows() can
influence the game-state. As such, they first need to acquire a
lock on the game-state before they can be called.
    
Currently, the main advantage is the time spend in Paint(), which
for non-OpenGL drivers can be a few milliseconds. For OpenGL this
is more like 0.05 milliseconds; in these instances this change
doesn't add any benefits for now.
    
This is an alternative to the former "draw-thread", which moved
the drawing in a thread for some OSes. It has similar performance
gain as this does, although this implementation allows for more
finer control over what suffers when the GameLoop takes too
long: drawing or the next GameLoop. For now they both suffer
equally.
```

This is the inverse of draw_thread, where drawing was done in a thread. Now the `GameLoop` is instead.

There is currently very little benefit from doing this with OpenGL. The time spend in `Paint`, the only function that is allowed to run in parallel with the `GameLoop` for now, takes on my machine ~0.05ms. That won't change anything, really.
For non-OpenGL, it is a bit different. `Paint` can easily take 5ms, so that means the simulation speed can go slightly higher.

If I load complex games, like ProGame5 or a game with many Timberwolf's trains, using threads does indeed increase the simulation speed (with something between 10% and 20%, so noticeable), when using non-OpenGL. With OpenGL, again, no difference is to be spotted (or expected).

This PR is mainly meant to allow future work. Things to think about:

- `UpdateWindows` now does everything while the game-state is locked. This is not required. Half-way through the viewport functions it can release the game-state lock, as it no longer will change the game-state. Sadly, this is not trivial to implement, but doing so would mean more time could be spend during the `GameLoop`, further improving the gameplay for complex games.
- During resource starvation, currently both the simulation and draw fps suffer. But this doesn't have to be that way. I experimented a bit with this, but it is trivial to let one or the other suffer more. In the end, if `GameLoop` goes over 30ms, draw fps will always suffer, but till that time simulation can suffer. Or, if we like, we can always prioritize simulation over draw fps, and for example have drawing being done at 10fps so the simulation can run at 1x. I have fiddled a bit with this, but it is really hard to find a good compromise. So l left it at equal starvation for now.
- The mouse is now non-responsive during `GameLoop`, but that is not needed. We can just keep the mouse reactive. This would also change how Modal windows work, for example. Already with this PR the `AcquireBlitterLock` doesn't do anything, and it is a small step to just remove it, and have a more reactive GUI.

## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

TrueBrain committed:
Remove: [Video] no longer draw in a thread

Drawing in a thread is a bit odd, and often leads to surprising
issues. For example, OpenGL would only allow it if you move the
full context to the thread. Which is not always easily done on
all OSes.
In general, the advise is to handle system events and drawing
from the main thread, and do everything else in other threads.
So, let's be more like other games.

Additionally, putting the drawing routine in a thread was only
done for a few targets.

Upcoming commit will move the GameLoop in a thread, which will
work for all targets.

TrueBrain committed:
Add: [Video] move GameLoop into its own thread

This allows drawing to happen while the GameLoop is doing an
iteration too.

Sadly, not much drawing currently can be done while the GameLoop
is running, as for example PollEvent() or UpdateWindows() can
influence the game-state. As such, they first need to acquire a
lock on the game-state before they can be called.

Currently, the main advantage is the time spend in Paint(), which
for non-OpenGL drivers can be a few milliseconds. For OpenGL this
is more like 0.05 milliseconds; in these instances this change
doesn't add any benefits for now.

This is an alternative to the former "draw-thread", which moved
the drawing in a thread for some OSes. It has similar performance
gain as this does, although this implementation allows for more
finer control over what suffers when the GameLoop takes too
long: drawing or the next GameLoop. For now they both suffer
equally.

frosch123 commented on src/video/video_driver.cpp:
```diff
@@ -15,69 +15,129 @@
+			 * got the lock. This avoids the game-thread to never yield.
+			 * See comment in the game-thread for more details. */
+			this->request_game_state_mutex = true;
+			std::lock_guard<std::mutex> lock(this->game_state_mutex);
```
I don't think this works. Example:
1. Tick(): request_game_state_mutex = true
2. Gameloop(): game_state_mutex.unlock()
3. Tick(): game_state_mutex.lock()
4. Gameloop(): request_game_state_mutex.load()
5. Tick(): notify_one(), but noone is waiting
6. Tick(): request_game_state_mutex = false
7. Gameloop(): waiting for condition_variable, but noone notifies

Instead you may try to use two mutexes: game_state_mutex, wait_mutex
* GameThead(): locks/unlocks wait_mutex between GameLoops. So this thread alternates between locking game_state_mutex and wait_mutex, never both.
* Tick(): Lock wait_mutex, then lock game_state_mutex.
So:
```
GameThread():
  for (;;) {
    {
      std::lock_guard lock(this->game_state_mutex);
      this->GameLoop();
    }
    {
      std::lock_guard lock(this->wait_mutex);
      std::this_thread::yield(); // just for fun
    }
  }

DrawThread():
  std::lock_guard lock1(this->wait_mutex); // easy to acquire while GameLoop is running
  std::lock_guard lock2(this->game_state_mutex); // easy to acquire because GameThread is blocked by wait_mutex
  ...
```

TrueBrain commented:
Sometimes, when switching game-modes, my mouse with OpenGL transforms into this:

![image](https://user-images.githubusercontent.com/1663690/109367913-b52dbb00-7897-11eb-8885-f4bf3cc6792f.png)

@michicc : happen to have any idea what could be causing that? It seems to be timing related, something to do with when GameLoop runs and when drawing is done.

michicc commented:
@TrueBrain I would assume that the call to `VideoDriver::ClearSystemSprites` is somehow mistimed.

TrueBrain commented:
> 
> 
> @TrueBrain I would assume that the call to `VideoDriver::ClearSystemSprites` is somehow mistimed.

W00p, tnx :) It was not this function, but the function populating the cache (OpenGL's `DrawMouseCursor`). This was done outside the lock on the game-state, but it was using `GetRawSprite` which is not valid to call when you don't own the lock. And indeed, `GameLoop` was rebuilding sprites, causing `DrawMouseCursor` to read random garbage instead.

Solution was to make a function that populated the cache for OpenGL, and call it while the game-state is still locked.

TrueBrain committed:
Codechange: remove the unused lock around Blitter

andythenorth commented:
No obvious issues from a basic check on Mac. 

It's 10-20% faster on ffwd for me than 1.11.0-beta-2 for same map area of a s specific savegame.  Might not be significant result, might just be noise in the system.

michicc committed:
Codechange: don't set the window position when changing blitter

There really is no need to make an extra call to the OS in
these cases.

TrueBrain committed:
Fix: ensure switching blitter happens in the main thread

This because video-drivers might need to make changes to their
context, which for most video-drivers has to be done in the same
thread as the window was created; main thread in our case.

michicc reviewed (dismissed):
Looks fine 😇 (after a rebase).
